### ISSUE TITLE
Axios returns ECONNREFUSED 127.0.0.1:80 but port is set to 8080

### ISSUE THREAD

**Issue by DragosRotaru** [axios/axios#3821]
#### Describe the bug

Axios returns ECONNREFUSED 127.0.0.1:80 but port is set to 8080

#### To Reproduce

While having an express server returning 200 on http://localhost:8080/ (and confirmed via browser), running the following returns the error provided (see additional context)

```js
axios({
  headers: { Accept: 'text/html, application/json, text/plain, */*' },
  proxy: undefined,
  url: 'http://localhost:8080',
  method: 'get'
});
```

#### Expected behavior

Axios sends an http request to the port specified in the url

#### Environment

Im indirectly using axios@0.21.1 (start-server-and-test -> wait-on -> axios@0.21.1).

 - Axios Version [0.21.1
 - Adapter HTTP
 - Node.js Version v12.8.0 and v15.8.0
 - OS: macOS 10.15.6

#### Additional context/Screenshots

Error:

```
Error: connect ECONNREFUSED 127.0.0.1:80
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1056:14) {
  errno: 'ECONNREFUSED',
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 80,
  config: {
    url: 'http://localhost:8080',
    method: 'get',
    headers: {
      Accept: 'text/html, application/json, text/plain, */*',
      'User-Agent': 'axios/0.21.1',
      host: 'localhost:8080'
    },
    transformRequest: [ [Function: transformRequest] ],
    transformResponse: [ [Function: transformResponse] ],
    timeout: 0,
    adapter: [Function: httpAdapter],
    xsrfCookieName: 'XSRF-TOKEN',
    xsrfHeaderName: 'X-XSRF-TOKEN',
    maxContentLength: -1,
    maxBodyLength: -1,
    httpsAgent: Agent {
      _events: [Object: null prototype],
      _eventsCount: 1,
      _maxListeners: undefined,
      defaultPort: 443,
      protocol: 'https:',
      options: [Object],
      requests: {},
      sockets: {},
      freeSockets: {},
      keepAliveMsecs: 1000,
      keepAlive: false,
      maxSockets: Infinity,
      maxFreeSockets: 256,
      maxCachedSessions: 100,
      _sessionCache: [Object]
    },
    validateStatus: [Function: validateStatus],
    data: undefined
  },
  request: Writable {
    _writableState: WritableState {
      objectMode: false,
      highWaterMark: 16384,
      finalCalled: false,
      needDrain: false,
      ending: false,
      ended: false,
      finished: false,
      destroyed: false,
      decodeStrings: true,
      defaultEncoding: 'utf8',
      length: 0,
      writing: false,
      corked: 0,
      sync: true,
      bufferProcessing: false,
      onwrite: [Function: bound onwrite],
      writecb: null,
      writelen: 0,
      bufferedRequest: null,
      lastBufferedRequest: null,
      pendingcb: 0,
      prefinished: false,
      errorEmitted: false,
      emitClose: true,
      autoDestroy: false,
      bufferedRequestCount: 0,
      corkedRequestsFree: [Object]
    },
    writable: true,
    _events: [Object: null prototype] {
      response: [Function: handleResponse],
      error: [Function: handleRequestError]
    },
    _eventsCount: 2,
    _maxListeners: undefined,
    _options: {
      maxRedirects: 21,
      maxBodyLength: 10485760,
      protocol: 'http:',
      path: 'http://localhost:8080/',
      method: 'GET',
      headers: [Object],
      agent: undefined,
      agents: [Object],
      auth: undefined,
      hostname: null,
      port: null,
      host: null,
      beforeRedirect: [Function: beforeRedirect],
      nativeProtocols: [Object],
      pathname: 'http://localhost:8080/'
    },
    _ended: true,
    _ending: true,
    _redirectCount: 0,
    _redirects: [],
    _requestBodyLength: 0,
    _requestBodyBuffers: [],
    _onNativeResponse: [Function],
    _currentRequest: ClientRequest {
      _events: [Object: null prototype],
      _eventsCount: 7,
      _maxListeners: undefined,
      outputData: [],
      outputSize: 0,
      writable: true,
      _last: true,
      chunkedEncoding: false,
      shouldKeepAlive: false,
      useChunkedEncodingByDefault: false,
      sendDate: false,
      _removedConnection: false,
      _removedContLen: false,
      _removedTE: false,
      _contentLength: 0,
      _hasBody: true,
      _trailer: '',
      finished: true,
      _headerSent: true,
      socket: [Socket],
      connection: [Socket],
      _header: 'GET http://localhost:8080/ HTTP/1.1\r\n' +
        'Accept: text/html, application/json, text/plain, */*\r\n' +
        'User-Agent: axios/0.21.1\r\n' +
        'host: localhost:8080\r\n' +
        'Connection: close\r\n' +
        '\r\n',
      _onPendingData: [Function: noopPendingOutput],
      agent: [Agent],
      socketPath: undefined,
      method: 'GET',
      path: 'http://localhost:8080/',
      _ended: false,
      res: null,
      aborted: false,
      timeoutCb: null,
      upgradeOrConnect: false,
      parser: null,
      maxHeadersCount: null,
      _redirectable: [Circular],
      [Symbol(isCorked)]: false,
      [Symbol(outHeadersKey)]: [Object: null prototype]
    },
    _currentUrl: 'http:http://localhost:8080/'
  },
  response: undefined,
  isAxiosError: true,
  toJSON: [Function: toJSON]
}```

**Comment by MrJmpl3** [comment]
I have same error but in Node, I remove `proxy: undefined`, and works.

I tried with `proxy: false` and this make error too

**Comment by Danil27** [comment]
Are you using docker?

**Comment by DragosRotaru** [comment]
No docker or any proxy involved.

**Comment by Danil27** [comment]
Try using the local ip of your computer instead of localhost. 
![image](https://user-images.githubusercontent.com/40491529/125284591-4d49af80-e322-11eb-8f81-49d8ec66804d.png)

**Comment by Danil27** [comment]
Wrap in try {} catch to view the error log

**Comment by alioergey** [comment]
i don't know about this, but it seems like the _currentURL is built wrong, `_currentUrl: 'http:http://localhost:8080/'`. It's doubled, can you try using it without `http://`

**Comment by qtv100291** [comment]
Same with me, when running absolute url with axios , ex: axios.get(http://localhost:8080/api/product), it works,
But when configuring axios instance, 
ex: axiosInstance = axios.create({baseUrl: http://localhost:8080/}) then run axiosInstance.get(/api/product), 
It doesn't work at all.

**Comment by karthikrocks** [comment]
I am using docker... But I get the error `Error: connect ECONNREFUSED` while connecting to my API... Any suggestions for this issue?

**Comment by Number16BusShelter** [comment]
is this a bug or is it the way it should work?

**Comment by alchemist-bounty** [comment]
is it really something related to proxy?

**Comment by Number16BusShelter** [comment]
> is it really something related to proxy?

It is probably not. I tried request without proxy, but the result was the same…

**Comment by alchemist-bounty** [comment]
I got the error while following the this link.
Error: connect ECONNREFUSED 127.0.0.1:80 TCPConnectWrap.afterConnect [as oncomplete] (node:net:1161:16)

https://auth0.com/docs/architecture-scenarios/spa-api/api-implementation-nodejs
Any idea on this?

**Comment by alchemist-bounty** [comment]
I found it. I specified wrong value in jwksUri field

**Comment by maximkoev** [comment]
> https://auth0.com/docs/architecture-scenarios/spa-api/api-implementation-nodejs

Hi threre, did you solve the issue, I faced this problem right now and have no idea what to do

**Comment by markg85** [comment]
I too have the very same issue as reported here.
I "think" i know "kindof" why.

Like the other examples here, i was using a `http://localhost:8080` or structured similarly URL.
This gave me a connection refused which is actually giving a hint:
```
Error: connect ECONNREFUSED ::1:8080
```
While i did **not in any place or location** specify IPv6, the error does indicate an IPv6 call, not v4!
In my case changing `localhost` to `127.0.0.1` did "fix" it.

Debugging this further with `ping localhost` showed me that magical `::1` too. It was an ipv6 ping.
I haven't quite figured it out yet but i'm guessing there is an error in the users `/etc/hosts` file where localhost is missing. I'll report back if i know more about that.

**Comment by christianirack** [comment]
If you're using environment variables, check the route.

In my case, I mistakenly commented my: //require('dotenv').config();

My route was underfined.
process.env.MY_ROUTE

**Comment by malickshola** [comment]
I recently have this issue with Axios and the solution was to check my URL. It seems that when you are calling an URL on local without **http://** or **https://** , Axios tried to connect to localhost.

**Comment by ocjadan** [comment]
Axios' RequestConfig has `baseUrl` and `url`.

For me, I was setting my server's url using `url` instead of `baseUrl` in `axios.create()` which was what was causing the error for me.

Making sure I used `baseUrl` when creating my `AxiosInstance` fixed the problem.

**Comment by Lightinway** [comment]
Thanks @malickshola !

Similar scenario here, I'm using Axios to make a request to my AWS load balancer DNS. Axios does not append the `http://` to the URL so I kept getting `200` but empty json response. Fixing my request URL fixed my issue.

**Comment by DragosRotaru** [comment]
Axios must adhere to the specification of a valid URL, that's just common sense. The user of an interface should not have to care about the implementation. `http://localhost:8080/` is a fully valid URL according to the RFC spec, and according to the Axios documentation.

As @aleyeeey  pointed out,  `_currentUrl: 'http:http://localhost:8080/'` is a good hint that Axios is doing something behind the scenes that it should not be doing.

The issue, probably lies in "proxy:undefined". The Axios documentation says the following:

> Use `false` to disable proxies, ignoring environment variables.

Therefore, I am going to close this issue on account that the request I have included is actually NOT adhering to the documentation. However, it is still suspect that defining `proxy:undefined` would lead to such an error, as the code should not have such a high specificity: null, false or undefined should be acceptable values, semantically speaking there is no reason NOT to accept `undefined`.

**Comment by jasonsaayman** [comment]
Hi 👋 

Please could you retry with the latest version and open a new issue should this error still be relevant?

Thanks

**Comment by gon777** [comment]
> I too have the very same issue as reported here. I "think" i know "kindof" why.
> 
> Like the other examples here, i was using a `http://localhost:8080` or structured similarly URL. This gave me a connection refused which is actually giving a hint:
> 
> ```
> Error: connect ECONNREFUSED ::1:8080
> ```
> 
> While i did **not in any place or location** specify IPv6, the error does indicate an IPv6 call, not v4! In my case changing `localhost` to `127.0.0.1` did "fix" it.
> 
> Debugging this further with `ping localhost` showed me that magical `::1` too. It was an ipv6 ping. I haven't quite figured it out yet but i'm guessing there is an error in the users `/etc/hosts` file where localhost is missing. I'll report back if i know more about that.
> 
> Edit. Don't know... It is something with the hosts and something with ipv6 but i don't know what triggers this. Using the ip instead does work but feels wrong.

i had exactly same issue, I checked my '/etc/hosts' and saw '::1 localhost', removing that line fixed my issue

**Comment by markg85** [comment]
> i had exactly same issue, I checked my '/etc/hosts' and saw '::1 localhost', removing that line fixed my issue

Don't do or even recommend that!

It's your localhost lookup. Your pc in IPv6 now doesn't know it's localhost anymore. Which is, frankly, probably fine as not much - if any - is using it. Still, it's very bad practice to delete that line.

**Comment by mudassir-jmi** [comment]
Server Error
Error: connect ECONNREFUSED 127.0.0.1:1337

This error happened while generating the page. Any console logs will be displayed in the terminal window.
Call Stack
<unknown>
error: connect ECONNREFUSED 127.0.0.1 (1337)
<unknown> (Error: connect ECONNREFUSED
127.0.0.1 (1337)
TCPConnectWrap.afterConnect [as oncomplete]
node:net **(1157:16)******

**Comment by joerush18** [comment]
127.0.0.1  add this instead of localhost.

**Comment by ConnorGDevelopment** [comment]
@joerush18 This solved the issue for me, thank you so much

**Comment by KoljaL** [comment]
> 127.0.0.1 add this instead of localhost.
> 
> http://localhost:8080 => http://127.0.0.1:8080

I had a similar issue with the same solution on Strapi.
Don't know why it works, but  it works...

**Comment by mudassir-jmi** [comment]
By default, Axios redirects us to 127.0.0.1:80 if it doesn't find the requested URL or HTTP method(GET/POST/PUT)
 check your URL if are also using Axios.

**Comment by sandeepkrai** [comment]
In my case, I was missing adding username and password in the .env file which I was referencing in the code.

**Comment by dulanjanabandara** [comment]
> I too have the very same issue as reported here. I "think" i know "kindof" why.
> 
> Like the other examples here, i was using a `http://localhost:8080` or structured similarly URL. This gave me a connection refused which is actually giving a hint:
> 
> ```
> Error: connect ECONNREFUSED ::1:8080
> ```
> 
> While i did **not in any place or location** specify IPv6, the error does indicate an IPv6 call, not v4! In my case changing `localhost` to `127.0.0.1` did "fix" it.
> 
> Debugging this further with `ping localhost` showed me that magical `::1` too. It was an ipv6 ping. I haven't quite figured it out yet but i'm guessing there is an error in the users `/etc/hosts` file where localhost is missing. I'll report back if i know more about that.
> 
> Edit. Don't know... It is something with the hosts and something with ipv6 but i don't know what triggers this. Using the ip instead does work but feels wrong.

This worked for me. An absolute gem.

**Comment by Arman091** [comment]
use your ip address shown in error in place of localhost

http:// 127.0.0.1:80 /api_url_to be followed

**Comment by sashiksu** [comment]
If you have this issue in the docker + Kubernetes setup, delete all local docker images and Kubernetes deployments and pods. For some reason, Kubernetes deployment is not taking a pull from the pushed remote docker image and this kind of Axios error throws.

**Comment by Kosisochukwu2** [comment]
Connection to Rusk Failed, some operations won't be available: error sending request for url (http://127.0.0.1:8080/): error trying to connect: tcp connect error: Connection refused (os error 111)

this happened while i was using putty

**Comment by arianemadureira** [comment]
I was using the docker, this error happen with me too, but I change the localhost to  my ipV4 it's works fine.
